File: 0234-palindrome-linked-list/0234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head, prev = None, nex = None):
        prev = None
        nex = None
        
        while head != None:
            nex = head.next
            head.next = prev
            prev = head
            head = nex
        
        return prev
    
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        
        if((head == None) or (head.next == None)):
            return True
        
        slow = head
        fast = head
        
        while((fast.next != None) and (fast.next.next != None)):
            slow = slow.next
            fast = fast.next.next
        
        slow.next = self.reverseList(slow.next)
        slow = slow.next
        
        
        while slow != None:
            if(head.val != slow.val):
                return False
            head = head.next
            slow = slow.next
            
        
        return True

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head){
        ListNode* Next = nullptr;
        ListNode* Prev = nullptr;
        
        while(head != nullptr){
            Next = head->next;
            head->next = Prev;
            Prev = head;
            head = Next;
        }
        return Prev;
    }
    
    bool isPalindrome(ListNode* head) {
        
        if(head == nullptr || head->next == nullptr){
            return true;
        }
        
        ListNode* slow = head;
        ListNode* fast = head;
        
        while(fast->next && fast->next->next){
            slow = slow->next;
            fast = fast->next->next;
        }
        
        slow->next = reverseList(slow->next);
        slow = slow->next;
        
        
        while(slow != nullptr){
            if(head->val != slow->val){
                return false;
            }
            slow = slow->next;
            head = head->next;
        }
        return true;
        
    }
};